src/Interfaces/FwupdManager.vala

```diff
@@ -0,0 +1,302 @@
+        } else {
+            device.releases = new List<Fwupd.Release> ();
+        }
+
+        return device;
+    }
+
+    private Fwupd.Release parse_release (GLib.HashTable<string, Variant> serialized_release) {
+        var release = new Fwupd.Release () {
+            icon = "security-high"
+        };
+
+        serialized_release.@foreach ((key, val) => {
+            switch (key) {
+                case "Filename":
+                    release.filename = val.get_string ();
+                    break;
+                case "Name":
+                    release.name = val.get_string ();
+                    break;
+                case "Summary":
+                    release.summary = val.get_string ();
+                    break;
+                case "Version":
+                    release.version = val.get_string ();
+                    break;
+                case "Description":
+                    release.description = val.get_string ()
+                    .replace ("<p>", "")
+                    .replace ("</p>", "\n\n")
+                    .replace ("<li>", " • ")
+                    .replace ("</li>", "\n")
+                    .replace ("<ul>", "")
+                    .replace ("</ul>", "\n")
+                    .replace ("<ol>", "") // TODO: add support for ordered lists
+                    .replace ("</ol>", "\n")
+                    .strip ();
+                    break;
+                case "Protocol":
+                    release.protocol = val.get_string ();
+                    break;
+                case "RemoteId":
+                    release.remote_id = val.get_string ();
+                    break;
+                case "AppstreamId":
+                    release.appstream_id = val.get_string ();
+                    break;
+                case "Checksum":
+                    release.checksum = val.get_string ();
+                    break;
+                case "Vendor":
+                    release.vendor = val.get_string ();
+                    break;
+                case "Size":
+                    release.size = val.get_uint64 ();
+                    break;
+                case "License":
+                    release.license = val.get_string ();
+                    break;
+                case "TrustFlags":
+                    release.flag = (Fwupd.ReleaseFlag) val.get_uint64 ();
+                    break;
+                case "InstallDuration":
+                    release.install_duration = val.get_uint32 ();
+                    break;
+                case "Uri":
+                    release.uri = val.get_string ();
+                    break;
+                default:
+                    break;
+            }
+        });
+
+        return release;
+    }
+
+    private string get_path (string uri) {
+        var parts = uri.split ("/");
+        string file_path = parts[parts.length - 1];
+        return Path.build_path (Path.DIR_SEPARATOR_S, Environment.get_tmp_dir (), file_path);
```

Comment:
You can use `Path.build_filename` and then you don't need to include `Path.DIR_SEPARATOR_S`